src/compiler.rs

```diff
@@ -97,12 +97,13 @@ pub fn compile(
             1,
         ),
         // Copy data
-        "Reshape" | "Dropout" | "Flatten" | "Squeeze" | "Softmax" => (
+        "Reshape" | "Dropout" | "Flatten" | "Squeeze" => (
             "endomorphism/copy.wgsl".to_string(),
             ceil(output_lengths[0], 16) as _,
             1,
             1,
         ),
+        "Softmax" => ("endomorphism/softmax.wgsl".to_string(), 1, 1, 1),
```

Comment:
So the `1, 1, 1` correspond to the number of threads you want to spawn. By doing this your only spawning 1 threads. I think the number of thread should be `output_lengths[0], 1, 1` although this will  only work if the axis of softmax is the first one.

Reply:
Hm, not sure I fully understand this yet. When 1 thread is spawned, I assume this means that only one instance of my shader will be running at any time (never in parallel). Then I guess this also means the shader will be invoked multiple times in series (as otherwise my current example would probably not work at all as only the first few elements would be calculated?).

It does seem a bit wasteful that each shader invocation is calculating max and sum each time but I guess the only way to fix that is to have multiple passes (which is more complex). 

Re axis, I think this is configurable for softmax, would have to add a check perhaps.

Reply:
Yep, so within the wgsl shader, you have the global_id index correspond to the thread of the shader. Each thread is independent of one another. If the threads are `1, 1, 1` only one thread is going to spawn and the only global_id that is going to spawn is the global_id = (0, 0, 0). If you want to compute the value for global_id from 0 to `output_lengths[0]` you have to specify it in advance. Therefore the `output_lengths[0], 1, 1`. 

I agree that it is quite wasteful, to do the summation in every thread. But, I'm ok with the current status. The thing is that as long as the softmax does not correspond to a large portion of the computation, like convolution for example. Doing optimisation will not result in visible gain.

Reply:
Thanks for the explanation, very helpful! 👍 

Now I'm just wondering why the current version even works at all. In my toy MNIST model there is a softmax outputting (10,), that should not be possible unless somehow the shader considers 10 elements at once (?) 🤔

Reply:
> Now I'm just wondering why the current version even works at all. In my toy MNIST model there is a softmax outputting (10,), that should not be possible unless somehow the shader considers 10 elements at once (?) 🤔

Never mind, I am simply looping from 0 to `i_lens[0]` of course, setting every element. 

So with multiple threads you would change the last loop to only set a few of the elements right? (Then i'm not sure having multiple threads will be faster because the only thing that is done in parallel is this last step, where as the sum/max are all the same because they consider all elements). It would be faster for cases with multiple axes however (and each axis getting their own thread).

Reply:
ah yes! Missed that part. Ok, well let's leave it as is.